system/include/emscripten/promise.h

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2023 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ */
+
+#pragma once
+
+#include <stdlib.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// An opaque handle to a JS Promise object.
+typedef struct em_promise_t {
+  void* handle;
+} em_promise_t;
+
+typedef enum em_promise_result_t {
+  EM_PROMISE_FULFILL,
+  EM_PROMISE_MATCH,
+  EM_PROMISE_MATCH_RELEASE,
+  EM_PROMISE_REJECT,
+} em_promise_result_t;
+
+// A callback passed to `emscripten_promise_then` to be invoked once a promise
+// is fulfilled or rejected. `data` is arbitrary user-provided data provided
+// when `emscripten_promise_then` is called to install the callback and `value`
+// is the value the promise was fulfilled or rejected with.
+//
+// The callback can signal how to resolve the new promise returned from
+// `emscripten_promise_then` via its return and by writing a new result to
+// outparam `result`. The behavior depends on the returned `em_promise_result_t`
+// value:
+//
+//  - `EM_PROMISE_FULFILL`: The new promise is fulfilled with the value written
+//    to `result` or NULL if no value is written.
+//
+//  - `EM_PROMISE_MATCH` or `EM_PROMISE_MATCH_RELEASE`: The callback must write
+//    a promise handle to `result` and the new promise is resolved to match the
+//    eventual state of that promise. `EM_PROMISE_MATCH_RELEASE` will also cause
+//    the written promise handle to be destroyed so that the user does not have
+//    to arrange for it to be destroyed after the callback is executed.
+//
+//  - `EM_PROMISE_REJECT`: The new promise is rejected with the reason written
+//    to `result` or NULL if no reason is written.
+typedef em_promise_result_t (*promise_callback_t)(void** result,
+                                                  void* data,
+                                                  void* value);
+
+// Create a new promise that can be explicitly resolved or rejected using
+// `emscripten_promise_resolve`. The returned promise handle must eventually be
+// freed with `emscripten_promise_destroy`.
+__attribute__((warn_unused_result)) em_promise_t
+emscripten_promise_create(void);
+
+// Release the resources associated with this promise. This must be called on
+// every promise handle created, whether by `emscripten_promise_create` or any
+// other function that returns a fresh promise, such as
+// `emscripten_promise_then`. It is fine to call `emscripten_promise_destroy` on
+// a promise handle before the promise is resolved; the configured callbacks
+// will still be called.
```

Comment:
How does that work? Don't the callbacks do something with the the handle in some way (which would no longer be valid)?

Reply:
No, the callbacks don't generally need the handle because they're installed directly on the underlying promise object. That underlying promise is kept alive by the JS runtime even if it removed from the handle map.